graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -442,13 +435,16 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
             # Make sure there's a marked location right before it for the optional Backtrack
             # to jump back to. Otherwise, the traversal could rewind to an old marked location
             # and might ignore entire stretches of applied filtering.
-            if context['marked_location_stack'][-1].num_traverses > 0:
+            # The only way there might not be a marked location here is if the current location
+            # has existing child locations that are not folds.
+            non_fold_child_locations = {
+                child_location
+                for child_location in query_metadata_table.get_child_locations(location)
+                if not isinstance(child_location, FoldScopeLocation)
+            }
+            if non_fold_child_locations:
```

Comment:
The reason I decided against the `any` short-circuit was because the code ended up looking a bit too complex for my taste. I wanted to give the intermediate value a name, for the sake of the reader's intuition, even if that meant slightly slower execution.